[docs][WIP] Workflow docs for python SDK (#562)
dapr/python-sdk: daprdocs/content/en/python-sdk-docs/python-workflow.md

{{% /alert %}}

Run the following command to start a workflow. 

{{< tabs "Linux/MacOS" "Windows">}}

{{% codetab %}}

```bash
todo
```

{{% /codetab %}}

{{% codetab %}}

```powershell
todo
```

{{% /codetab %}}

{{< /tabs >}}

If successful, you should see a response like the following: 

```json
todo
```

Send an HTTP request to get the status of the workflow that was started:

```bash
todo
```

The workflow is designed to take several seconds to complete. If the workflow hasn't completed when you issue the HTTP request, you'll see the following JSON response (formatted for readability) with workflow status as `RUNNING`:

```json
todo
```

Once the workflow has completed running, you should see the following output, indicating that it has reached the `COMPLETED` status:

```json
todo
```

When the workflow has completed, the stdout of the workflow app should look like:

```log
todo
```

## Error Handling

The Dapr Python SDK will pass through errors that it receives from the Dapr runtime. In the case of an ETag mismatch, the Dapr runtime will return `StatusCode.ABORTED`.

**Example**

```

> == APP == State store has successfully saved value_1 with key_1 as key

Reviewer: These outputs are not currently correct and are only present in the python PR since the output at this time is not known. Once the example app is created the expected output will be updated

---
type: docs
title: "Getting started with the Dapr Workflow Python SDK"
linkTitle: "Workflow"
weight: 30000
description: How to get up and running with workflows using the Dapr Python SDK
---

> Let’s create a Dapr workflow and invoke it using the console. With the [provided hello world workflow example](https://github.com/dapr/python-sdk/tree/master/examples/demo_workflow), you will:

Reviewer: I'd remove the "hello world" from the description here

== APP == Instance Successfully Purged

== APP == start_resp exampleInstanceID

== APP == Hi Counter!
== APP == New counter value is: 1112!

== APP == Hi Counter!
== APP == New counter value is: 1122!

== APP == Get response from hello_world_wf after terminate call: Terminated
== APP == Instance Successfully Purged
```

## What happened?

When you ran `dapr run`, the Dapr client:
1. Registered the workflow (`hello_world_wf`) and its actvity (`hello_act`)
1. Started the workflow engine

```python
def main():
    with DaprClient() as d:
        host = settings.DAPR_RUNTIME_HOST
        port = settings.DAPR_GRPC_PORT
        workflowRuntime = WorkflowRuntime(host, port)
        workflowRuntime = WorkflowRuntime()
        workflowRuntime.register_workflow(hello_world_wf)
        workflowRuntime.register_activity(hello_act)
        workflowRuntime.start()

        print("==========Start Counter Increase as per Input:==========")
        start_resp = d.start_workflow(instance_id=instanceId, workflow_component=workflowComponent,
                        workflow_name=workflowName, input=inputData, workflow_options=workflowOptions)
        print(f"start_resp {start_resp.instance_id}")
```

> Dapr then paused and resumed the workflow:

Reviewer 2: ```suggestion
Pause and the resume the workflow:
```

```python
       # Pause
        d.pause_workflow(instance_id=instanceId, workflow_component=workflowComponent)
        getResponse = d.get_workflow(instance_id=instanceId, workflow_component=workflowComponent)
        print(f"Get response from {workflowName} after pause call: {getResponse.runtime_status}")

        # Resume
        d.resume_workflow(instance_id=instanceId, workflow_component=workflowComponent)
        getResponse = d.get_workflow(instance_id=instanceId, workflow_component=workflowComponent)
        print(f"Get response from {workflowName} after resume call: {getResponse.runtime_status}")
```

Once the workflow resumed, Dapr raised a workflow event and printed the new counter value:

```python
        # Raise event
        d.raise_workflow_event(instance_id=instanceId, workflow_component=workflowComponent,
                    event_name=eventName, event_data=eventData)
```

> To clear out the workflow state from your state store, Dapr purged the workflow:

Reviewer 2: ```suggestion
To clear out the workflow state from your state store, purge the workflow:
```

```python
        # Purge
        d.purge_workflow(instance_id=instanceId, workflow_component=workflowComponent)
        try:
            getResponse = d.get_workflow(instance_id=instanceId, workflow_component=workflowComponent)
        except DaprInternalError as err:
            if nonExistentIDError in err._message:
                print("Instance Successfully Purged")
```

The sample then demonstrated terminating a workflow by:
- Starting a new workflow using the same `instanceId` as the purged workflow.

> - Terminating the workflow and purging before shutting down the workflow.

Reviewer 2: ```suggestion
- Terminating the workflow and purging again before shutting down the workflow.
```
